in-progress/7228-unified-bbjs-interface.md

```diff
@@ -0,0 +1,247 @@
+
+Since we don't have plans on reworking bb as a long-running native process now, we are limited to communicate with native bb via its CLI, so we'll shoot for a coarse API. This should also be simpler for users to work with. Eventually, if we implement the long-running native bb process, we can provide a more fine-grained API, and reimplement the one in this doc via multiple calls to it.
+
+For the design, we gather "inspiration" from the native CLI, the bb.js CLI, the noir-js-bb-backend, and the Aztec bb-prover.
+
+## Interface and implementation
+
+We propose the following interface, to be implemented via the native CLI and wasm as described below.
+
+### Types
+
+#### `Circuit`
+
+Circuit bytecode in ACIR format.
+
+#### `Witness`
+
+Binary witness as outputted by the ACVM simulator.
+
+#### ~~`PathOrData`~~
+
+Either a path to the filesystem or the actual contents. This would allow us to propagate paths to disk from the bb.js interface into the native bb, instead of having to read a file into memory from bb.js just to conform to the shared interface, and then write again to disk to call the native bb CLI.
+
+:question: We discarded this type for simplicity reasons, but there are reasons to keep it. Even after we discard the native CLI, we may still want to just point native bb to disk as a means to share data between bb.js and bb, especially if we cache more aggressively.
+
+#### `Flavor`
+
+Specifies the proving system for the proof. Not all methods may be supported for all flavors, either because of limitations (eg AVM is not supported in wasm) or because it doesn't make sense (eg there is a single AVM verification key). Valid values:
+
+- `UltraPlonk`
+- `UltraHonk`
+- `MegaHonk`
+- `ClientIVC`
+- `AVM`
+
+:question: I collected these values from scanning the bb CLI interface. Are they correct? Should we only target a subset of them? Is "flavor" the correct term? Should we spell it properly as "flavour"?
+
+### Interface
+
+#### `prove(circuit: Circuit, witness: Witness, flavor: Flavor): { proof: Buffer }`
+
+Proves a given circuit with the given witness as input. Returns the proof in raw binary format.
+
+Native implementation:
+
+- `bb prove | prove_ultra_honk | prove_mega_honk | avm_prove | client_ivc_prove_output_all` (there is no `client_ivc_prove`)
+
+Wasm implementation (`AVM` not supported)
+
+- `acirGetCircuitSizes`
+- `commonInitSlabAllocator`
+- `srsInitSrs`
+- `acirNewAcirComposer`
+- `acirInitProvingKey`
+- `acirCreateProof | acirProveUltraHonk | acirProveAndVerifyMegaHonk | acirFoldAndVerifyProgramStack` (there is no `acirProveMegaHonk`, hence the `acirProveAndVerifyMegaHonk`)
+
+#### `proveForRecursion(circuit: Circuit, witness: Witness, flavor: Flavor): { proof: Buffer, proofAsFields: Fields, vk: Buffer, vkAsFields: Fields }`
```

Comment:
Currently Noir embeds in the `circuit` whether it should be proven/verified using a recursion-friendly hash function.

Reply:
Good to know. Is this embedded in the bytecode of the circuit, or somewhere else in the artifact?

Reply:
https://github.com/noir-lang/noir/blob/6936a52bb4de2170b6e4c41afd9cf4327d4fbeb1/acvm-repo/acir/src/circuit/mod.rs#L75

It's currently a boolean flag in the `Circuit` struct so not in the "bytecode" if we class that as just the ACIR opcodes but in the binary blob currently sent to barretenberg.

May move to `Program` in future: https://github.com/noir-lang/noir/issues/5364